sdk/core/azure-core/src/exception.cpp

```diff
@@ -28,4 +30,48 @@ namespace Azure { namespace Core {
     RawResponse = std::move(rawResponse);
   }
 
+  RequestFailedException::RequestFailedException(
+      std::unique_ptr<Azure::Core::Http::RawResponse>& rawResponse)
+      : std::runtime_error(getRawResponseField(rawResponse, "message"))
+  {
+    auto& headers = rawResponse->GetHeaders();
+
+    StatusCode = rawResponse->GetStatusCode();
+    ErrorCode = getRawResponseField(rawResponse, "code");
+    StatusCode = rawResponse->GetStatusCode();
+    ReasonPhrase = rawResponse->GetReasonPhrase();
+    RequestId = HttpShared::GetHeaderOrEmptyString(headers, HttpShared::MsRequestId);
+    ClientRequestId = HttpShared::GetHeaderOrEmptyString(headers, HttpShared::MsClientRequestId);
+    Message = this->what();
+  }
+
+  std::string RequestFailedException::getRawResponseField(
+      std::unique_ptr<Azure::Core::Http::RawResponse>& rawResponse,
+      std::string fieldName)
+  {
+    auto& headers = rawResponse->GetHeaders();
+    std::string contentType = HttpShared::GetHeaderOrEmptyString(headers, HttpShared::ContentType);
+    std::vector<uint8_t> bodyBuffer = rawResponse->GetBody();
+    std::string result;
+
+    if (contentType.find("json") != std::string::npos)
+    {
+      auto jsonParser = Azure::Core::Json::_internal::json::parse(bodyBuffer);
+      auto error = jsonParser.find("error");
+      if (error != jsonParser.end() && error.value().contains(fieldName))
+      {
+        result = error.value()[fieldName].get<std::string>();
+      }
+      else if (fieldName == "message")
+      {
+        result = std::string(bodyBuffer.begin(), bodyBuffer.end());
+      }
+    }
```

Comment:
I didn't move this to Core because the structure of the Response payload from each service for an error can be different. Storage can use XML, Keyvault uses Json, other service can use Json but with other field names.  It would be hard to maintain a one only method to cover all future possibilities. If this is the case, the factory method should take an strategy callback which defines how to parse the payload.